File: src/main/java/dev/kikugie/elytratrims/common/recipe/ETGlowRecipe.java
package dev.kikugie.elytratrims.common.recipe;

import dev.kikugie.elytratrims.common.ETServer;
import dev.kikugie.elytratrims.common.access.FeatureAccess;
import net.minecraft.inventory.Inventory;
import net.minecraft.item.ItemStack;
import net.minecraft.item.Items;
import net.minecraft.recipe.RecipeSerializer;
import net.minecraft.recipe.SpecialRecipeSerializer;
import net.minecraft.recipe.book.CraftingRecipeCategory;
import net.minecraft.util.Identifier;

public class ETGlowRecipe extends DelegatedRecipe {
    public static RecipeSerializer<ETGlowRecipe> SERIALIZER = new SpecialRecipeSerializer<>(ETGlowRecipe::new);
    public ETGlowRecipe(Identifier id, CraftingRecipeCategory category) {
        super(id, category);
    }

    @Override
    boolean matches(Inventory inventory) {
        int item = 0;
        int sac = 0;
        for (int slot = 0; slot < inventory.size(); slot++) {
            ItemStack stack = inventory.getStack(slot);
            if (ETServer.isProbablyElytra(stack.getItem())) item++;
            else if (stack.getItem() == Items.GLOW_INK_SAC) sac++;
            else if (!stack.isEmpty()) return false;
            if (item > 1 || sac > 1) return false;
        }
        return item == 1 && sac == 1;
    }

    @Override
    ItemStack craft(Inventory inventory) {
        ItemStack item = ItemStack.EMPTY;
        for (int slot = 0; slot < inventory.size(); slot++) {
            ItemStack stack = inventory.getStack(slot);
            if (!ETServer.isProbablyElytra(stack.getItem())) continue;
            item = stack.copy();
        }
        FeatureAccess.setGlow(item, true);
        return item;
    }

    @Override
    public boolean fits(int width, int height) {
        return width * height >= 2;
    }

    @Override
    public RecipeSerializer<?> getSerializer() {
        return null;
    }
}

File: src/main/kotlin/dev/kikugie/elytratrims/common/recipe/ETGlowRecipe.kt
package dev.kikugie.elytratrims.common.recipe

import dev.kikugie.elytratrims.api.ElytraTrimsAPI
import dev.kikugie.elytratrims.common.access.FeatureAccess.addGlow
import net.minecraft.item.ItemStack
import net.minecraft.item.Items
import net.minecraft.recipe.RecipeSerializer
import net.minecraft.util.Identifier

class ETGlowRecipe(id: Identifier) : DelegatedRecipe(id, SAMPLE)  {
    override fun matches(input: Stacks): Boolean {
        var item = 0
        var sac = 0
        input.forEach {
            if (ElytraTrimsAPI.isProbablyElytra(it)) item++
            else if (it.item == Items.GLOW_INK_SAC) sac++
            else if (!it.isEmpty) return false
            if (item > 1 || sac > 1) return false
        }
        return item == 1 && sac == 1
    }

    override fun craft(input: Stacks): ItemStack {
        val elytra = input.firstItemCopy(ElytraTrimsAPI::isProbablyElytra) ?: return ItemStack.EMPTY
        elytra.addGlow()
        return elytra
    }

    override fun fits(width: Int, height: Int): Boolean = width * height >= 2

    companion object {
        val SAMPLE = ItemStack(Items.ELYTRA).apply { addGlow() }
    }
}